### DISCUSSION TITLE
K6 사용법

### DISCUSSION THREAD

**Discussion by kyum-q** [woowacourse-teams/2024-code-zap#786] [기록해봐요]
# 설치 방법
# 실행 방법
## 간단한 실행 결과
# 실행 결과 값들의 의미

> 참고
> https://subeen-lab.tistory.com/128
> https://k6.io/docs/get-started/running-k6/

**Comment by kyum-q** [comment]
# K6 설치

## 윈도우 😊

### 1. chocolatey 설치
```bash
 Set-ExecutionPolicy Bypass -Scope Process -Force; [System.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bor 3072; iex ((New-Object System.Net.WebClient).DownloadString('https://community.chocolatey.org/install.ps1'))
```

### 2. choco 설치 확인
```bash
choco
```

### 3. k6 설치
```bash
choco install k6
```

## 맥

### 1. k6 설치
```bash
brew install k6
```

**Reply by HoeSeong123** [reply]
역시 맥이 좋아~~

**Comment by kyum-q** [comment]
# k6 실행

## 테스트

### 1. 폴더 하나 만들어서 javascript 파일 생성

> 파일명 아무거나 상관 없음, 현재는 `test.js`로 한다

```js
import http from 'k6/http';
import { sleep } from 'k6';

export default function () {
    http.get('http://test.k6.io'); // http.get('확인하고 싶은 경로');
    sleep(1);
}
```

### 2. bash에서 해당 경로로 이동한 후 실행
```bash
k6 run test.js
```

#### 실행 결과

![image](https://github.com/user-attachments/assets/94bb2ace-8a7c-43cc-98a4-3a926aad7feb)

---

## 추가적인 코드 설정

### html로 결과 확인

#### 1. 파일 상단에 해당 코드 추가
```js
import { htmlReport } from "https://raw.githubusercontent.com/benc-uk/k6-reporter/main/dist/bundle.js";
```

#### 2. 파일 하단에 해당 코드 추가
```js
export function handleSummary(data) {
    return {
        "summary.html" : htmlReport(data)
    }
}
```

### 결과

> 실행하면 test.js와 동일한 경로에 html 생성됨

![image](https://github.com/user-attachments/assets/614220fd-b440-4444-be2c-ed3ee6033646)

---

## 지속 시간 및 사용자 설정 가능

### 옵션 설정 : 해당 코드 추가
```js
export const options = {
    stages: [
// 지속 시간(duration): 30초 가상 사용자(target): 20명
// 각 가상 사용자가 1초에 한 번씩 요청을 보내므로: 20명 × 30초 = 600개의 요청
        { duration: '30s', target: 20 },
    ],
};
```

### 여러 옵션을 추가할 수도 있음
```js
export const options = {
    stages: [
        { duration: '30s', target: 20 },
        { duration: '1m30s', target: 10 },
        { duration: '20s', target: 0 },
    ],
};
```

**Comment by kyum-q** [comment]
# 실행 결과 값들의 의미

 - data_received : [counter] 데이터 전달받은 양
 - data_sent : [counter] 데이터 전달 양 
 - http_req_blocked : [trend] 요청 시작 전 차단된 시간 (TCP connection slot을 기다리는)
 - http_req_connecting : [trend] 원격 호스트에 대한 TCP 연결을 설정하는데 소요된 시간
 - http_req_duration : [trend] 요청의 총 시간 = http_req_sending + http_req_waiting + http_req_receiving
 - http_req_failed : [trend] setResponseCallbak에 따른 요청 실패
 - http_req_receiving : [trend] 원격 호스트로부터 응답 데이터를 수신하는데 소요된 시간
 - http_req_sending : [trend] 원격 호스트에 데이터를 보내는데 소요된 시간
 - http_req_tls_handshaking : [trend] 원격 호스트와의 handshaking TLS 세션에 소요된 시간
 - http_req_waiting : [trend] 원격 호스트로부터의 응답을 대기하는 데 소요된 시간 (TTFB ; time to first byte)
 - http_reqs : [counter] 총 얼마나 많은 http requests를 k6에서 생성했는지 횟수
 - iteration_duration [trend] default/main function의 전체 반복을 한 번 완료하는데 소요된 시간
 - iterations : [counter] 테스트에서 Vu가 js 스크립트를 실행한 총 횟수
 - vus : [gauge] 현재 활성화된 사용자 유저
 - vus_max : [gauge] 가능한 최대 가상 사용자 수 (로드 레벨을 확장할 때 성능에 영향을 미치지 않도록 vu 리소스가 미리 할당됨)

**Comment by kyum-q** [comment]
# 로컬 서버로 우리 서비스 테스트도 해봄 (1)

`http://localhost:8080/templates?memberId=2` api에 대해 테스트

## 현재 데이터 수

### 템플릿 26000개
![image](https://github.com/user-attachments/assets/6813d183-6539-4d9b-9ea2-b1e9faa85be8)

### 성능평가 테스트 스크립트
```js
import http from 'k6/http';
import { sleep } from 'k6';
import { htmlReport } from "https://raw.githubusercontent.com/benc-uk/k6-reporter/main/dist/bundle.js";

export const options = {
    stages: [
// 지속 시간(duration): 30초 가상 사용자(target): 20명
// 각 가상 사용자가 1초에 한 번씩 요청을 보내므로: 20명 × 30초 = 600개의 요청
        { duration: '30s', target: 20 },
        { duration: '1m30s', target: 10 },
        { duration: '20s', target: 0 },
    ],
};

export default function () {
    http.get('http://localhost:8080/templates?memberId=2');
    sleep(1);
}

export function handleSummary(data) {
    return {
        "summary.html" : htmlReport(data)
    }
}
```

### 실행

```bash
k6 run test.js
```

### 결과

![image](https://github.com/user-attachments/assets/64d7137c-e2c2-4869-944a-d6762fbe31ea)

![image](https://github.com/user-attachments/assets/1ef20638-6848-4bdf-bafd-9d82f7f8735e)

## 의미 분석
이 K6 로드 테스트 결과를 분석해보면 다음과 같습니다:

### 1. **총 요청 수 (Total Requests):**
   - **1541건**의 요청이 처리되었습니다. 이는 K6 스크립트가 테스트 기간 동안 서버에 1541번의 요청을 보냈다는 것을 의미합니다.

### 2. **실패한 요청 (Failed Requests):**
   - **0건**의 요청이 실패했습니다. 모든 요청이 성공적으로 처리되었음을 나타냅니다.

### 3. **Threshold 위반 (Breached Thresholds):**
   - **0건**의 Threshold 위반이 발생하지 않았습니다. 이 값은 미리 설정한 임계값을 넘는 요청이 없었다는 것을 의미합니다.

### 4. **실패한 체크 (Failed Checks):**
   - **0건**의 체크 실패가 발생하지 않았습니다. 이는 모든 검증(Validation)이 통과했음을 나타냅니다.

### 5. **요청 처리 시간 (Request Metrics):**
   - **http_req_duration (전체 요청 소요 시간):** 
     - 평균 응답 시간: **171.95ms**
     - 최댓값: **700.23ms**
     - 중앙값: **162.18ms**
     - 90번째 백분위수: **220.67ms**
     - 95번째 백분위수: **239.16ms**
     - 요청의 대부분이 220ms 이내에 완료된 것을 확인할 수 있습니다.
   
   - **http_req_waiting (대기 시간):**
     - 평균 대기 시간: **171.58ms** 
     - 대기 시간은 서버에서 응답이 준비될 때까지의 시간을 나타내며, 전체 응답 시간과 거의 비슷합니다. 이는 네트워크 지연보다 서버 응답 시간이 성능에 더 영향을 미친다는 것을 시사합니다.

   - **http_req_connecting (연결 시간):** 
     - 평균 연결 시간: **0.03ms** 
     - 이 값이 매우 짧으므로 네트워크 연결 성능에는 큰 문제가 없음을 나타냅니다.

   - **iteration_duration (각 실행 시간):** 
     - 평균 반복 시간: **1179.34ms**
     - 이는 각 사용자 시나리오 실행에 소요된 시간을 나타냅니다. 스크립트에서 각 실행마다 1초의 `sleep`을 사용했으므로 이 시간은 예상대로 나옵니다.

### **전반적인 분석:**
- **응답 시간은 안정적**이며, 요청의 대부분이 200ms 이하로 완료되었습니다. 이는 상당히 빠른 응답 시간으로, 서버가 부하 상황에서도 적절히 응답하고 있음을 보여줍니다.
- **실패한 요청이 없다는 점**에서 시스템 안정성이 뛰어남을 알 수 있습니다.
- **대기 시간이 전체 요청 시간과 비슷**하므로, 서버 내부 처리 성능을 좀 더 최적화하면 성능이 더 개선될 여지가 있을 수 있습니다.

이 결과를 통해 시스템이 현재 설정된 부하 수준에서 적절하게 처리하고 있으며, 성능 문제는 크게 없는 것으로 보입니다.

**Comment by kyum-q** [comment]
# 로컬 서버로 우리 서비스 테스트도 해봄 (2)
`http://localhost:8080/templates?keyword=a` api에 대해 테스트

## 현재 데이터 수

### 템플릿 26000개
![image](https://github.com/user-attachments/assets/6813d183-6539-4d9b-9ea2-b1e9faa85be8)

### 성능평가 테스트 스크립트
```js
import http from 'k6/http';
import { sleep } from 'k6';
import { htmlReport } from "https://raw.githubusercontent.com/benc-uk/k6-reporter/main/dist/bundle.js";

export const options = {
    stages: [
        { duration: '30s', target: 20 },
        { duration: '1m30s', target: 10 },
        { duration: '20s', target: 0 },
    ],
};

export default function () {
    http.get('http://localhost:8080/templates?keyword=a');
    sleep(1);
}

export function handleSummary(data) {
    return {
        "summary.html" : htmlReport(data)
    }
}
```

### 실행

```bash
k6 run test.js
```

### 결과

![image](https://github.com/user-attachments/assets/76df2d3a-ecb9-45e2-acfd-8d677a76b63a)

![image](https://github.com/user-attachments/assets/773592ad-2d1a-43e1-8f98-7e47f9ede1b5)


## 의미 분석

이 K6 로드 테스트 결과를 분석해보면 다음과 같습니다:

### 1. **총 요청 수 (Total Requests):**
   - **1541건**의 요청이 처리되었습니다. 이는 K6 스크립트가 테스트 기간 동안 서버에 1541번의 요청을 보냈다는 것을 의미합니다.

### 2. **실패한 요청 (Failed Requests):**
   - **0건**의 요청이 실패했습니다. 모든 요청이 성공적으로 처리되었음을 나타냅니다.

### 3. **Threshold 위반 (Breached Thresholds):**
   - **0건**의 Threshold 위반이 발생하지 않았습니다. 이 값은 미리 설정한 임계값을 넘는 요청이 없었다는 것을 의미합니다.

### 4. **실패한 체크 (Failed Checks):**
   - **0건**의 체크 실패가 발생하지 않았습니다. 이는 모든 검증(Validation)이 통과했음을 나타냅니다.

### 5. **요청 처리 시간 (Request Metrics):**
   - **http_req_duration (전체 요청 소요 시간):** 
     - 평균 응답 시간: **171.95ms**
     - 최댓값: **700.23ms**
     - 중앙값: **162.18ms**
     - 90번째 백분위수: **220.67ms**
     - 95번째 백분위수: **239.16ms**
     - 요청의 대부분이 220ms 이내에 완료된 것을 확인할 수 있습니다.
   
   - **http_req_waiting (대기 시간):**
     - 평균 대기 시간: **171.58ms** 
     - 대기 시간은 서버에서 응답이 준비될 때까지의 시간을 나타내며, 전체 응답 시간과 거의 비슷합니다. 이는 네트워크 지연보다 서버 응답 시간이 성능에 더 영향을 미친다는 것을 시사합니다.

   - **http_req_connecting (연결 시간):** 
     - 평균 연결 시간: **0.03ms** 
     - 이 값이 매우 짧으므로 네트워크 연결 성능에는 큰 문제가 없음을 나타냅니다.

   - **iteration_duration (각 실행 시간):** 
     - 평균 반복 시간: **1179.34ms**
     - 이는 각 사용자 시나리오 실행에 소요된 시간을 나타냅니다. 스크립트에서 각 실행마다 1초의 `sleep`을 사용했으므로 이 시간은 예상대로 나옵니다.

### **전반적인 분석:**
- **응답 시간은 안정적**이며, 요청의 대부분이 200ms 이하로 완료되었습니다. 이는 상당히 빠른 응답 시간으로, 서버가 부하 상황에서도 적절히 응답하고 있음을 보여줍니다.
- **실패한 요청이 없다는 점**에서 시스템 안정성이 뛰어남을 알 수 있습니다.
- **대기 시간이 전체 요청 시간과 비슷**하므로, 서버 내부 처리 성능을 좀 더 최적화하면 성능이 더 개선될 여지가 있을 수 있습니다.

이 결과를 통해 시스템이 현재 설정된 부하 수준에서 적절하게 처리하고 있으며, 성능 문제는 크게 없는 것으로 보입니다.

**Comment by HoeSeong123** [comment]
집이라서 실제로 테스트 해보지는 못했지만 아래 링크에서처럼 그라파나를 이용해서 GUI로도 쉽게 확인할 수 있는 것 같네요.
[K6 성능 테스트](https://velog.io/@hooni_/K6-%EC%84%B1%EB%8A%A5-%ED%85%8C%EC%8A%A4%ED%8A%B8)

**Comment by HoeSeong123** [comment]
다음과 같이 시나리오도 작성할 수 있습니다. 
```
import http from "k6/http";
import { check } from "k6";

export const options = {
  stages: [  
    { duration: '2m', target: 5 }, 
    { duration: '10m',target: 5 },
    { duration: '3m', target: 10 },
    { duration: '30m',target: 10 },
    { duration: '3m', target: 0 },
  ],

  thresholds: { 
    http_req_duration: ['p(95)<100'],
  }
};

export default function () {
  const response = http.get("http://localhost:8080/posts");
  check(response, {
    "success": (res) => res.status === 200
  });
}
```
해석해보면 다음과 같습니다.

2분 동안 5명으로 유저를 증가시키고
10분 동안 5명의 유저를 유지시킵니다.
그리고 3분 동안 10명으로 유저를 늘립니다.
피크 시간은 30분 동안 진행되고 10명의 유저를 유지시킵니다.
그리고 3분간 0명의 유저로 줄입니다.

또, 95%가 100ms 안에 응답을 받는 것을 목표로 합니다.

**Comment by zeus6768** [comment]
# 시나리오1: 가장 부하가 많은 상황

가능한 상황: 7기 프리코스 대상 홍보 직후, 7기 프리코스 2차 미션 발표 직후, 프리코스 최종코테

# 시나리오2: 평상시

실제로 하루 100건이므로 TPS로 계산하면 0.001TPS 🥲

그렇게 테스트할 순 없을 것 같으니 조금 올려볼까?

**Reply by zangsu** [reply]
# 테스트 방법

누적 사람 수와 시간을 정한다. 
이를 기준으로 10분동안의 동시 접속자 수를 구한다. 
동시 접속자 수가 LCP 시간을 만족시킬 수 있도록 성능을 개선한다.

## 부하가 많은 상황

### 1. 템플릿 목록 조회 (구경가기)

배경:  7기 프리코스 대상 홍보 직후
최대 사람 수: ((디스코드 7기 사람 수 * 0.6) + 100) * 0.5
시간: 30분 (5분동안 올리고, 25분 유지)
예산: LCP 2초 (Gist 는 1.6초)

### 2. 템플릿 키워드 검색

배경: 7기 프리코스 대상 홍보 직후
사람 수: ((디스코드 7기 사람 수 * 0.6) + 100) * 0.5 -> ((디스코드 7기 사람 수 * 0.2) + 20) * 0.5
시간: 30분 (5분동안 올리고, 25분 유지)
예산: LCP 3초 (Gist 는 1.6초)

(자주 사용할 것 같은 키워드를 리스트로 설정해서 랜덤으로 사용한다.)
-> 스프링, Java, 자바, 우테코, 자바스크립트, 코틀린, 안드로이드, 검증

### 3. 템플릿 저장

배경: 2차 발표 직후 합격자들의 최종 코딩테스트 준비
사람 수: 300명
시간: 30분  (5분동안 올리고 25분동안 유지)
예산: LCP 0.8초
1. 사용자의 템플릿 리스트를 조회한다. -> 0.5초
2. 사용자의 템플릿 상세조회한다. -> 0.3초

## 평상시

사람 수: 50명
시간: 30분 (5분 올리고 25분동안 유지)
예산: 
1. 로그인 한다. -> 0.5초
3. 본인 템플릿 리스트를 확인한다. -> 0.5초
4. 카테고리 조건을 추가해서 조회한다. -> 0.7초
5. 태그 조건을 추가해서 조회한다. -> 0.7초
6. 템플릿 중 하나를 상세조회한다. -> 0.3초
7. 템플릿을 하나 생성한다. -> 1초

**Comment by kyum-q** [comment]
## Standard built-in metrics

k6는 테스트에서 사용하는 프로토콜에 관계없이 항상 다음 메트릭을 수집합니다

| Metric Name        | Type    | Description                                                             |
| ------------------ | ------- | ----------------------------------------------------------------------- |
| checks             | Rate    | 성공률                                                                     |
| data_received      | Counter | 수신된 데이터의 양                                                              |
| data_sent          | Counter | 전송된 데이터의 양                                                              |
| dropped_iterations | Counter | VU 부족(도착률 실행자) 또는 시간 부족(반복 기반 실행자에서 만료된 maxDuration)으로 인해 시작되지 않은 반복 횟수 |
| iteration_duration | Trend   | 전체 반복을 완료하는 데 걸리는 시간                                                    |
| iterations         | Counter | VU가 JS 스크립트(함수)를 실행하는 총 횟수                                              |
| vus                | Gauge   | 현재 활성 가상 사용자 수                                                          |
| vus_max            | Gauge   | 가능한 최대 가상 사용자 수                                                         |
## HTTP-specific built-in metrics

이러한 메트릭은 테스트가 HTTP 요청을 할 때만 생성됩니다.

| Metric Name              | Type    | Description                                                                                                            |
| ------------------------ | ------- | ---------------------------------------------------------------------------------------------------------------------- |
| http_req_blocked         | Trend   | 요청을 시작하기 전에 block되는 데 소요된 시간(사용 가능한 TCP 연결 슬롯을 기다리는 데 소요된 시간) `float`                                                  |
| http_req_connecting      | Trend   | 원격 호스트에 TCP 연결을 설정하는 데 소요된 시간  `float`                                                                                 |
| http_req_duration        | Trend   | HTTP 요청 응답 시간<br> `http_req_sending + http_req_waiting + http_req_receiving` `float`                                   |
| http_req_failed          | Rate    | [setResponseCallback](https://grafana.com/docs/k6/latest/javascript-api/k6-http/set-response-callback/) 에 따른 실패한 요청 비율 |
| http_req_receiving       | Trend   | 원격 호스트로부터 응답 데이터를 받는 데 소요된 시간 `float`                                                                                  |
| http_req_sending         | Trend   | 원격 호스트로 데이터를 전송하는 데 소요된 시간 `float`                                                                                     |
| http_req_tls_handshaking | Trend   | 원격 호스트와 TLS 세션을 핸드셰이크하는 데 소요된 시간                                                                                       |
| http_req_waiting         | Trend   | 원격 호스트로부터 응답을 기다리는 데 소요된 시간 `float`                                                                                    |
| http_reqs                | Counter | k6가 생성한 총 HTTP 요청 수                                                                                                    |
